src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala

```diff
@@ -2,33 +2,41 @@ package mesosphere.marathon.upgrade
 
 import akka.actor.{ Actor, ActorLogging, Cancellable }
 import akka.event.EventStream
+import mesosphere.marathon.Protos.MarathonTask
 import mesosphere.marathon.TaskUpgradeCanceledException
 import mesosphere.marathon.core.launchqueue.LaunchQueue
 import mesosphere.marathon.event.{ HealthStatusChanged, MesosStatusUpdateEvent }
+import mesosphere.marathon.health.{ Health, HealthCheckManager }
 import mesosphere.marathon.state.AppDefinition
 import mesosphere.marathon.tasks.TaskTracker
 import mesosphere.marathon.upgrade.TaskReplaceActor._
 import org.apache.mesos.Protos.TaskID
+import org.apache.mesos.Protos.TaskState._
 import org.apache.mesos.SchedulerDriver
+import org.slf4j.LoggerFactory
 
 import scala.collection.mutable
-import scala.concurrent.Promise
+import scala.concurrent.{ Await, Promise }
 import scala.concurrent.duration._
 
 class TaskReplaceActor(
     driver: SchedulerDriver,
     taskQueue: LaunchQueue,
     taskTracker: TaskTracker,
+    healthCheckManager: HealthCheckManager,
     eventBus: EventStream,
     app: AppDefinition,
     promise: Promise[Unit]) extends Actor with ActorLogging {
   import context.dispatcher
 
-  val tasksToKill = taskTracker.get(app.id)
+  private[this] val log = LoggerFactory.getLogger(getClass.getName)
+
+  val newTasks = taskTracker.get(app.id).filter(_.getVersion == app.version.toString)
+  val tasksToKill = taskTracker.get(app.id) -- newTasks
```

Comment:
Ahh ok. It looks like you made `TaskTracker.getTaskMap` private so I went with`Set[TaskID]` for this operation. It might be cleaner to get the map though... Let me know if you think I should make it public and use the map.